src/Microsoft.Azure.WebJobs.Host/Scale/DefaultHostProcessMonitor.cs

```diff
@@ -0,0 +1,259 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+// Some of the logic in this file was moved from https://github.com/Azure/azure-functions-host/blob/852eed9ceef6ef56b431428a8eb31f1bd9c97f3b/src/WebJobs.Script/Scale/HostPerformanceManager.cs 
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    internal class DefaultHostProcessMonitor : IHostProcessMonitor, IDisposable
+    {
+        internal const string CpuLimitName = "CPU";
+        internal const string MemoryLimitName = "Memory";
+
+        private const int MinSampleCount = 5;
+        private readonly long _maxMemoryThresholdBytes;
+        private readonly ProcessMonitor _processMonitor;
+        private readonly List<ProcessMonitor> _childProcessMonitors = new List<ProcessMonitor>();
+        private readonly IOptions<ConcurrencyOptions> _options;
+        private readonly object _syncLock = new object();
+
+        private bool _disposed;
+
+        public DefaultHostProcessMonitor(IOptions<ConcurrencyOptions> options, ProcessMonitor processMonitor = null)
+        {
+            _options = options;
```

Comment:
I recommend you add a null-check here just to be extra safe and make it clear to the reader of this code what your expectations are. Since this is internal code, I'd say you probably don't need to bother with this if you decide to enable nullable reference types for this file.

```suggestion
            _options = options ?? throw new ArgumentNullException(nameof(options));
```

Reply:
Adding null checks here, and in many of the other new types I've added that were missing them